File: game/src/starfield.py
"""Enhanced starfield with parallax scrolling and nebula effects"""
import pygame
import random
import math
from typing import List, Tuple
from .config import SCREEN_WIDTH, SCREEN_HEIGHT

class Star:
    """A single star in the starfield"""
    def __init__(self, x: float, y: float, size: float, brightness: int, depth: float):
        self.x = x
        self.y = y
        self.size = size
        self.brightness = brightness
        self.depth = depth  # 0.0-1.0, lower = farther away
        self.twinkle_offset = random.random() * 6.28  # Random phase for twinkling

class Starfield:
    """Enhanced starfield with parallax scrolling"""
    
    def __init__(self):
        # Multiple layers for parallax effect
        self.layers: List[List[Star]] = []
        self.num_layers = 3
        
        # Layer properties (depth, density, size_range, brightness_range, parallax_factor)
        self.layer_configs = [
            # Far layer - small, dim, slow parallax
            (0.2, 150, (1, 1), (80, 120), 0.2),
            # Mid layer - medium size and brightness
            (0.5, 100, (1, 2), (120, 180), 0.5),
            # Near layer - larger, brighter, fast parallax
            (0.8, 50, (2, 3), (180, 255), 1.0),
        ]
        
        # Nebula particles for background atmosphere
        self.nebula_particles: List[Tuple[float, float, float, Tuple[int, int, int, int]]] = []
        
        self._generate_starfield()
        self._generate_nebula()
    
    def _generate_starfield(self):
        """Generate stars for all layers"""
        for depth, count, size_range, brightness_range, _ in self.layer_configs:
            layer = []
            for _ in range(count):
                x = random.randint(0, SCREEN_WIDTH * 3)
                y = random.randint(0, SCREEN_HEIGHT * 3)
                size = random.uniform(*size_range)
                brightness = random.randint(*brightness_range)
                star = Star(x, y, size, brightness, depth)
                layer.append(star)
            self.layers.append(layer)
    
    def _generate_nebula(self):
        """Generate nebula particles for atmospheric effect"""
        num_nebula = 20
        colors = [
            (50, 0, 80, 30),    # Purple
            (0, 30, 80, 30),    # Blue
            (80, 0, 40, 30),    # Magenta
            (0, 50, 60, 30),    # Cyan
        ]
        
        for _ in range(num_nebula):
            x = random.randint(0, SCREEN_WIDTH * 4)
            y = random.randint(0, SCREEN_HEIGHT * 4)
            radius = random.randint(80, 200)
            color = random.choice(colors)
            self.nebula_particles.append((x, y, radius, color))
    
    def render(self, screen: pygame.Surface, camera_x: float, camera_y: float, time: float):
        """Render the starfield with parallax scrolling"""
        # Render nebula first (background)
        self._render_nebula(screen, camera_x, camera_y)
        
        # Render star layers with parallax
        for layer_idx, layer in enumerate(self.layers):
            _, _, size_range, _, parallax_factor = self.layer_configs[layer_idx]
            
            for star in layer:
                # Apply parallax based on depth
                parallax_x = camera_x * parallax_factor
                parallax_y = camera_y * parallax_factor
                
                # Calculate screen position with wrapping
                world_width = SCREEN_WIDTH * 3
                world_height = SCREEN_HEIGHT * 3
                
                screen_x = (star.x - parallax_x) % world_width
                screen_y = (star.y - parallax_y) % world_height
                
                # Only draw if on screen (with margin)
                if -50 < screen_x < SCREEN_WIDTH + 50 and -50 < screen_y < SCREEN_HEIGHT + 50:
                    # Add subtle twinkling effect
                    twinkle = math.sin(time * 2 + star.twinkle_offset) * 0.2 + 0.8
                    brightness = int(star.brightness * twinkle)
                    brightness = max(0, min(255, brightness))
                    
                    color = (brightness, brightness, brightness)
                    
                    if star.size > 1:
                        # Draw larger stars as circles
                        pygame.draw.circle(screen, color, (int(screen_x), int(screen_y)), int(star.size))
                    else:
                        # Draw small stars as single pixels
                        if 0 <= screen_x < SCREEN_WIDTH and 0 <= screen_y < SCREEN_HEIGHT:
                            screen.set_at((int(screen_x), int(screen_y)), color)
    
    def _render_nebula(self, screen: pygame.Surface, camera_x: float, camera_y: float):
        """Render nebula clouds in the background"""
        # Create a temporary surface for nebula with alpha
        nebula_surface = pygame.Surface((SCREEN_WIDTH, SCREEN_HEIGHT), pygame.SRCALPHA)
        
        for x, y, radius, color in self.nebula_particles:
            # Very slow parallax for nebula (0.1x camera movement)
            parallax_x = camera_x * 0.1
            parallax_y = camera_y * 0.1
            
            world_width = SCREEN_WIDTH * 4
            world_height = SCREEN_HEIGHT * 4
            
            screen_x = (x - parallax_x) % world_width
            screen_y = (y - parallax_y) % world_height
            
            # Draw nebula as soft gradients
            if -radius < screen_x < SCREEN_WIDTH + radius and -radius < screen_y < SCREEN_HEIGHT + radius:
                # Create radial gradient effect
                steps = 5
                for i in range(steps):
                    r = radius * (1 - i / steps)
                    alpha = int(color[3] * (1 - i / steps))
                    current_color = (color[0], color[1], color[2], alpha)
                    pygame.draw.circle(nebula_surface, current_color, (int(screen_x), int(screen_y)), int(r))
        
        # Blit nebula surface onto main screen
        screen.blit(nebula_surface, (0, 0))

File: Starfield.cs
using Microsoft.Xna.Framework;
using Microsoft.Xna.Framework.Graphics;
using System;
using System.Collections.Generic;

namespace Subspace;

/// <summary>
/// A single star in the starfield
/// </summary>
public class Star
{
    public float X { get; set; }
    public float Y { get; set; }
    public float Size { get; set; }
    public int Brightness { get; set; }
    public float Depth { get; set; }
    public float TwinkleOffset { get; set; }

    public Star(float x, float y, float size, int brightness, float depth)
    {
        X = x;
        Y = y;
        Size = size;
        Brightness = brightness;
        Depth = depth;
        TwinkleOffset = (float)(new Random().NextDouble() * Math.PI * 2);
    }
}

/// <summary>
/// Enhanced starfield with parallax scrolling
/// </summary>
public class Starfield
{
    private const float TWINKLE_SPEED = 2.0f;
    private const float TWINKLE_AMPLITUDE = 0.2f;
    private const float TWINKLE_BASE = 0.8f;

    private List<List<Star>> layers = new List<List<Star>>();
    private Random random = new Random();

    // Layer properties (depth, density, minSize, maxSize, minBrightness, maxBrightness, parallaxFactor)
    private readonly (float depth, int count, float minSize, float maxSize, int minBrightness, int maxBrightness, float parallaxFactor)[] layerConfigs = 
    {
        (0.2f, 150, 1f, 1f, 80, 120, 0.2f),    // Far layer
        (0.5f, 100, 1f, 2f, 120, 180, 0.5f),   // Mid layer
        (0.8f, 50, 2f, 3f, 180, 255, 1.0f),    // Near layer
    };

    public Starfield()
    {
        GenerateStarfield();
    }

    private void GenerateStarfield()
    {
        foreach (var config in layerConfigs)
        {
            var layer = new List<Star>();
            for (int i = 0; i < config.count; i++)
            {
                float x = (float)random.NextDouble() * Config.SCREEN_WIDTH * 3;
                float y = (float)random.NextDouble() * Config.SCREEN_HEIGHT * 3;
                float size = config.minSize + (float)random.NextDouble() * (config.maxSize - config.minSize);
                int brightness = config.minBrightness + random.Next(config.maxBrightness - config.minBrightness);
                var star = new Star(x, y, size, brightness, config.depth);
                layer.Add(star);
            }
            layers.Add(layer);
        }
    }

    public void Render(SpriteBatch spriteBatch, Texture2D pixelTexture, float cameraX, float cameraY, float gameTime)
    {
        for (int layerIdx = 0; layerIdx < layers.Count; layerIdx++)
        {
            var layer = layers[layerIdx];
            var config = layerConfigs[layerIdx];

            foreach (var star in layer)
            {
                // Apply parallax based on depth
                float parallaxX = cameraX * config.parallaxFactor;
                float parallaxY = cameraY * config.parallaxFactor;

                // Calculate screen position with wrapping
                float worldWidth = Config.SCREEN_WIDTH * 3;
                float worldHeight = Config.SCREEN_HEIGHT * 3;

                float screenX = (star.X - parallaxX) % worldWidth;
                float screenY = (star.Y - parallaxY) % worldHeight;

                if (screenX < 0) screenX += worldWidth;
                if (screenY < 0) screenY += worldHeight;

                // Only draw if on screen (with margin)
                if (screenX >= -50 && screenX <= Config.SCREEN_WIDTH + 50 && 
                    screenY >= -50 && screenY <= Config.SCREEN_HEIGHT + 50)
                {
                    // Add subtle twinkling effect
                    float twinkle = (float)Math.Sin(gameTime * TWINKLE_SPEED + star.TwinkleOffset) * TWINKLE_AMPLITUDE + TWINKLE_BASE;
                    int brightness = (int)(star.Brightness * twinkle);
                    brightness = Math.Clamp(brightness, 0, 255);

                    Color color = new Color(brightness, brightness, brightness);

                    if (star.Size > 1)
                    {
                        // Draw larger stars as circles
                        DrawCircle(spriteBatch, pixelTexture, (int)screenX, (int)screenY, (int)star.Size, color);
                    }
                    else
                    {
                        // Draw small stars as single pixels
                        spriteBatch.Draw(pixelTexture, new Rectangle((int)screenX, (int)screenY, 1, 1), color);
                    }
                }
            }
        }
    }

    private void DrawCircle(SpriteBatch spriteBatch, Texture2D texture, int centerX, int centerY, int radius, Color color)
    {
        for (int y = -radius; y <= radius; y++)
        {
            for (int x = -radius; x <= radius; x++)
            {
                if (x * x + y * y <= radius * radius)
                {
                    spriteBatch.Draw(texture, new Rectangle(centerX + x, centerY + y, 1, 1), color);
                }
            }
        }
    }
}
